rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,306 @@
+tags. A new tags map will be added where the values are lists of optional strings, matching the
+semantics given above.
+
+```protobuf
+message Metric {
+  …
+  map<string, string> tags_v1 = 3;
+  message Values {
+    message Value {
+      optional string value = 1;
+    }
+    repeated Value values = 1;
+  }
+  map<string, Values> tags_v2 = 20;
+  …
+}
+```
+
+Similarly, the native JSON encoding for tags is a simple JSON object containing string-to-string
+pairs. This will be enhanced to also allow the `null` value for bare tags, and arrays of either
+strings or `null` for tags with more than one value.
+
+```json
+{
+  "tags": {
+    "single_value":"value",
+    "bare_tag":null,
+    "multi_valued_tag":["value1","value2"],
+    "complex_tag":["value3",null]
+  }
+}
+```
+
+#### Scripting
+
+##### VRL
+
+VRL exposes the tags as a simple map of string values in `.tags`. This will be changed to exposing
+the values as arrays containing either string values or `null`, with the latter representing a bare
+tag. Assignment of a single value to a tag will overwrite all other values for the tag and trigger a
+deprecation warning. Deleting a tag name will remove all tag values with that name. As the values
+will be arrays, these can be used to add or remove individual values. Any tag consisting of an empty
+list will be removed. Since VRL has a "compile" phase, where type information is determined
+statically, we will use this information to generate warnings and errors about obsolete assignments
+at start up time. Since tags cannot have multiple identical values, modifications that result in
+duplicate values will also cause the duplicate to be dropped and a warning issued.
+
+Examples:
+
+```vrl
+.tags.single_value = "value"
+.tags.bare_tag = null
+.tags.multi_valued_tag = ["value1", "value2"]
+.tags.complex_tag = ["value3", null]
+.tags.modified = push(.tags.modified, "value4")
+.tags.modified = filter(.tags.modified) -> |_, v| { v != "remove" }
+```
+
+##### Lua
+
+The Lua transform also exposes tags as a simple map of string values in `event.metric.tags`. Similar
+to VRL, this will be changed to have arrays containing either string values or `null`. Since the
+script is under complete control of the data between starting the script and emitting data, we
+cannot issue warnings about how assignments to the tags are made. Instead, the resulting tags data
+will be parsed using the same algorithm as the native JSON encoding above.
+
+#### `log_to_metric` Transform
+
+The `log_to_metric` transform is configured with a set of tags for the output metrics. Existing
+configurations with single value assignments will continue to work as they are currently
+written. Support will be added for assignments of arrays to produce multi-valued tags and null
+values to produce bare tags. As TOML does not support null values, the latter will be only be
+supported in YAML and JSON configuration.
```

Comment:
I'd really support dropping TOML support 💭

Reply:
Of course you would.